Automattic/jetpack
# Social: Move share limits code to Publicize package

#### Changes proposed in this Pull Request:

This PR moves the existing share limits code into the Publicize package, and initialises it in Jetpack Social.

Using it in Jetpack will be handled in a follow-up PR.

* Move the share limits code to the Publicize package.
* Introduce a new build step for the Publicize package.

#### Other information:


#### Jetpack product discussion


1202583289493814-as-1202966425123040/f

#### Does this pull request change what data or activity we track or use?


No.

#### Testing instructions:






_The testing instructions are very similar to https://github.com/Automattic/jetpack/pull/25661 and https://github.com/Automattic/jetpack/pull/26039_

* Make sure tests pass: `jetpack test plugins/social php` and `jetpack test packages/publicize php`
* Enable the Jetpack Social plugin.
* Add at least two connections to your testing site.
* Make a new post, and check that the sharing screen looks like the first image.
* Enable sharing limits for your site (first line in test plan of D85993-code)

For the block editor:
* Make a new post, and check that you can see the upgrade notice (second image).
* Test that connections are disabled by default when there's more connections than shares left, and that you can't select more connections than allowed.
  * To mock the number of shares left, you can add `$shares['shares_remaining'] = 1;` to `initial_state()` and `$shares_remaining = 1;` to `set_up_sharing_limits()` in `class-publicize-base.php:1502`
* Finally, enable the Jetpack plugin, and make sure the screen looks like the first screenshot again

Then, classic editor:
* Deactivate Jetpack again.
* Activate the Classic Editor plugin.
* Start writing a new post.
* Make sure the notice shows with the number of shares left and the upgrade link.
* To mock the number of shares left, you can add `$shares['shares_remaining'] = 1;` to `initial_state()` and `$shares_remaining = 1;` to `set_up_sharing_limits()` in `class-publicize-base.php:1502`
* Start a new post again, and check the following:
    * All connections are disabled.
    * You can only check one connection at a time.
<table>
<tr>
	<td>Sharing Limits Disabled 
	<td>Sharing Limits Enabled
	<td>More connections than shares left
	<td>More connections than shares left, and all remaining shares have been used
<tr>
	<td><img width="278" alt="image" src="https://user-images.githubusercontent.com/1713699/184976111-03be1d83-04b1-449c-bdea-583b8281a292.png">
	<td><img width="278" alt="image" src="https://user-images.githubusercontent.com/1713699/184976450-e0c32894-9eb3-4402-a5ea-826f367c2d25.png">
	<td><img width="280" alt="image" src="https://user-images.githubusercontent.com/1713699/184976463-949ade72-0653-4c15-95a3-e20fa62945a4.png">
	<td><img width="280" alt="image" src="https://user-images.githubusercontent.com/1713699/184976475-d15d5279-9561-4d0f-b536-f7b34e431524.png">
</table>
